gcc/rust/typecheck/rust-tyty-bounds.cc

```diff
@@ -103,46 +106,76 @@ TypeBoundPredicate::TypeBoundPredicate (
     reference (trait_reference.get_mappings ().get_defid ()), locus (locus),
-    args (HIR::GenericArgs::create_empty ()), error_flag (false)
-{}
+    error_flag (false)
+{
+  // we setup a dummy implict self argument
+  SubstitutionArg placeholder_self (&get_substs ().front (), nullptr);
+  used_arguments.get_mappings ().push_back (placeholder_self);
+}
 
 TypeBoundPredicate::TypeBoundPredicate (
   DefId reference, std::vector<SubstitutionParamMapping> substitutions,
   Location locus)
   : SubstitutionRef (std::move (substitutions),
 		     SubstitutionArgumentMappings::error ()),
-    reference (reference), locus (locus),
-    args (HIR::GenericArgs::create_empty ()), error_flag (false)
-{}
+    reference (reference), locus (locus), error_flag (false)
+{
+  // we setup a dummy implict self argument
+  SubstitutionArg placeholder_self (&get_substs ().front (), nullptr);
+  used_arguments.get_mappings ().push_back (placeholder_self);
+}
 
 TypeBoundPredicate::TypeBoundPredicate (const TypeBoundPredicate &other)
-  : SubstitutionRef ({}, other.used_arguments), reference (other.reference),
-    locus (other.locus), args (other.args), error_flag (other.error_flag)
+  : SubstitutionRef ({}, SubstitutionArgumentMappings::error ()),
+    reference (other.reference), locus (other.locus),
+    error_flag (other.error_flag)
 {
   substitutions.clear ();
-  if (!other.is_error ())
+
+  for (const auto &p : other.get_substs ())
+    substitutions.push_back (p.clone ());
+
+  std::vector<SubstitutionArg> mappings;
+  for (size_t i = 0; i < other.used_arguments.get_mappings ().size (); i++)
     {
-      for (const auto &p : other.get_substs ())
-	substitutions.push_back (p.clone ());
+      const SubstitutionArg &oa = other.used_arguments.get_mappings ().at (i);
+      TyTy::BaseType *argument
+	= oa.get_tyty () == nullptr ? nullptr : oa.get_tyty ()->clone ();
+      SubstitutionArg arg (&substitutions.at (i), argument);
+      mappings.push_back (std::move (arg));
     }
+
+  used_arguments
+    = SubstitutionArgumentMappings (mappings,
+				    other.used_arguments.get_locus ());
 }
 
 TypeBoundPredicate &
 TypeBoundPredicate::operator= (const TypeBoundPredicate &other)
 {
   reference = other.reference;
   locus = other.locus;
-  args = other.args;
   error_flag = other.error_flag;
-  used_arguments = other.used_arguments;
+  used_arguments = SubstitutionArgumentMappings::error ();
 
   substitutions.clear ();
-  if (!other.is_error ())
+  for (const auto &p : other.get_substs ())
+    substitutions.push_back (p.clone ());
+
+  std::vector<SubstitutionArg> mappings;
+  for (size_t i = 0; i < other.used_arguments.get_mappings ().size (); i++)
     {
-      for (const auto &p : other.get_substs ())
-	substitutions.push_back (p.clone ());
+      const SubstitutionArg &oa = other.used_arguments.get_mappings ().at (i);
+      TyTy::BaseType *argument
+	= oa.get_tyty () == nullptr ? nullptr : oa.get_tyty ()->clone ();
```

Comment:
This is the second time this ternary is used, I think it might be worth it to refactor it in a function or method so that it's clearer to use and easier to understand?

Reply:
yeah, I might be able to update the copy constructor on the SubstitutionArg to make this safer. There is an awkward case as Traits have an implicit Self type-parameter in order to do proper compatibility checking when a trait is implemented.

Reply:
fixed